File: sdk/python/stacyvm/sandbox.py
"""StacyVM Sandbox — represents a single sandbox instance."""

from __future__ import annotations

from typing import Iterator

import httpx

from stacyvm.exceptions import handle_response
from stacyvm.models import ExecResult
from stacyvm.streaming import StreamChunk, iter_ndjson


class Sandbox:
    """A StacyVM sandbox instance. Supports context manager for auto-cleanup."""

    def __init__(self, http: httpx.Client, sandbox_id: str, info: dict | None = None):
        self._http = http
        self.id = sandbox_id
        self._info = info or {}

    @property
    def state(self) -> str:
        return self._info.get("state", "unknown")

    @property
    def provider(self) -> str:
        return self._info.get("provider", "")

    @property
    def image(self) -> str:
        return self._info.get("image", "")

    def exec(
        self,
        command: str,
        args: list[str] | None = None,
        mode: str | None = None,
        env: dict[str, str] | None = None,
        workdir: str | None = None,
        timeout: str | None = None,
    ) -> ExecResult:
        """Execute a command in the sandbox."""
        body: dict = {"command": command}
        if args:
            body["args"] = args
        if mode:
            body["mode"] = mode
        if env:
            body["env"] = env
        if workdir:
            body["workdir"] = workdir
        if timeout:
            body["timeout"] = timeout

        resp = self._http.post(f"/api/v1/sandboxes/{self.id}/exec", json=body)
        handle_response(resp)
        data = resp.json()
        return ExecResult(
            exit_code=data["exit_code"],
            stdout=data["stdout"],
            stderr=data["stderr"],
            duration=data.get("duration", ""),
        )

    def exec_stream(
        self,
        command: str,
        args: list[str] | None = None,
        mode: str | None = None,
        env: dict[str, str] | None = None,
        workdir: str | None = None,
    ) -> Iterator[StreamChunk]:
        """Execute a command and stream output chunks."""
        body: dict = {"command": command, "stream": True}
        if args:
            body["args"] = args
        if mode:
            body["mode"] = mode
        if env:
            body["env"] = env
        if workdir:
            body["workdir"] = workdir

        with self._http.stream("POST", f"/api/v1/sandboxes/{self.id}/exec", json=body) as resp:
            handle_response(resp)
            yield from iter_ndjson(resp)

    def write_file(self, path: str, content: str, mode: str | None = None) -> None:
        """Write a file to the sandbox."""
        body: dict = {"path": path, "content": content}
        if mode:
            body["mode"] = mode
        resp = self._http.post(f"/api/v1/sandboxes/{self.id}/files", json=body)
        handle_response(resp)

    def read_file(self, path: str) -> str:
        """Read a file from the sandbox."""
        resp = self._http.get(f"/api/v1/sandboxes/{self.id}/files", params={"path": path})
        handle_response(resp)
        return resp.text

    def list_files(self, path: str = "/") -> list[dict]:
        """List files in a sandbox directory."""
        resp = self._http.get(f"/api/v1/sandboxes/{self.id}/files/list", params={"path": path})
        handle_response(resp)
        return resp.json()

    def delete_file(self, path: str, recursive: bool = False) -> None:
        """Delete a file or directory from the sandbox."""
        params = {"path": path}
        if recursive:
            params["recursive"] = "true"
        resp = self._http.delete(f"/api/v1/sandboxes/{self.id}/files", params=params)
        handle_response(resp)

    def move_file(self, old_path: str, new_path: str) -> None:
        """Move/rename a file in the sandbox."""
        resp = self._http.post(
            f"/api/v1/sandboxes/{self.id}/files/move",
            json={"old_path": old_path, "new_path": new_path},
        )
        handle_response(resp)

    def chmod_file(self, path: str, mode: str) -> None:
        """Change file permissions in the sandbox."""
        resp = self._http.post(
            f"/api/v1/sandboxes/{self.id}/files/chmod",
            json={"path": path, "mode": mode},
        )
        handle_response(resp)

    def stat_file(self, path: str) -> dict:
        """Get file info for a single file in the sandbox."""
        resp = self._http.get(
            f"/api/v1/sandboxes/{self.id}/files/stat", params={"path": path}
        )
        handle_response(resp)
        return resp.json()

    def glob_files(self, pattern: str) -> list[str]:
        """Return paths matching a glob pattern in the sandbox."""
        resp = self._http.get(
            f"/api/v1/sandboxes/{self.id}/files/glob", params={"pattern": pattern}
        )
        handle_response(resp)
        return resp.json()

    def extend_ttl(self, ttl: str = "30m") -> None:
        """Extend this sandbox's TTL. New expiry is calculated from now, not from the current expiry."""
        resp = self._http.post(f"/api/v1/sandboxes/{self.id}/extend", json={"ttl": ttl})
        handle_response(resp)
        self.refresh()

    def destroy(self) -> None:
        """Destroy this sandbox."""
        resp = self._http.delete(f"/api/v1/sandboxes/{self.id}")
        handle_response(resp)

    def refresh(self) -> None:
        """Refresh sandbox info from the server."""
        resp = self._http.get(f"/api/v1/sandboxes/{self.id}")
        handle_response(resp)
        self._info = resp.json()

    def get_preview_url(self, port: int) -> str:
        """Get the live preview URL for a given port."""
        domain = self._info.get("preview_domain", "localhost")
        return f"http://{port}-{self.id}.{domain}"

    def __enter__(self):
        return self

    def __exit__(self, *args):
        try:
            self.destroy()
        except Exception:
            pass

    def __repr__(self) -> str:
        return f"Sandbox(id={self.id!r}, state={self.state!r})"

File: sdk/js/src/sandbox.ts
/**
 * Sandbox class -- represents a single running sandbox instance.
 *
 * A {@link Sandbox} is never constructed directly.  Instead, obtain one via
 * {@link Client.spawn}, {@link Client.get}, or
 * {@link TemplateManager.spawn}.
 *
 * @module sandbox
 */

import { handleResponse } from "./errors.js";
import { parseNDJSON } from "./streaming.js";
import type {
  ExecOptions,
  ExecResult,
  FileInfo,
  SandboxInfo,
  SandboxState,
  StreamChunk,
} from "./types.js";

/**
 * A StacyVM sandbox instance.
 *
 * Provides methods to execute commands, manage files, and control the
 * lifecycle of a sandbox running on the StacyVM server.
 *
 * @example
 * ```ts
 * const sandbox = await client.spawn({ image: "alpine:latest" });
 *
 * const result = await sandbox.exec("echo hello");
 * console.log(result.stdout); // "hello\n"
 *
 * await sandbox.writeFile("/tmp/greeting.txt", "Hello, world!");
 * const content = await sandbox.readFile("/tmp/greeting.txt");
 *
 * await sandbox.destroy();
 * ```
 */
export class Sandbox {
  /** Base URL of the StacyVM API (no trailing slash). */
  private readonly _baseUrl: string;
  /** Default headers sent with every request. */
  private readonly _headers: Record<string, string>;
  /** Request timeout in milliseconds. */
  private readonly _timeout: number;
  /** Cached sandbox info from the server. */
  private _info: SandboxInfo;

  /**
   * @internal -- Use {@link Client.spawn} or {@link Client.get} instead.
   */
  constructor(
    baseUrl: string,
    headers: Record<string, string>,
    timeout: number,
    info: SandboxInfo,
  ) {
    this._baseUrl = baseUrl;
    this._headers = headers;
    this._timeout = timeout;
    this._info = info;
  }

  // -----------------------------------------------------------------------
  // Read-only properties
  // -----------------------------------------------------------------------

  /** Unique sandbox identifier (e.g. `"sb-a1b2c3d4"`). */
  get id(): string {
    return this._info.id;
  }

  /** Current state of the sandbox. */
  get state(): SandboxState {
    return this._info.state;
  }

  /** Container / VM image this sandbox was created from. */
  get image(): string {
    return this._info.image;
  }

  /** Provider running this sandbox. */
  get provider(): string {
    return this._info.provider;
  }

  /** Memory allocated in megabytes. */
  get memoryMb(): number {
    return this._info.memory_mb;
  }

  /** Number of virtual CPUs. */
  get vcpus(): number {
    return this._info.vcpus;
  }

  /** Full sandbox info snapshot. */
  get info(): SandboxInfo {
    return { ...this._info };
  }

  /**
   * Get the live preview URL for a given port.
   *
   * @param port - The internal port to expose (e.g. 3000).
   * @returns The full HTTP URL for the preview.
   *
   * @example
   * ```ts
   * const url = sandbox.getPreviewUrl(3000);
   * console.log(`Preview: ${url}`);
   * ```
   */
  getPreviewUrl(port: number): string {
    const domain = this._info.preview_domain || "localhost";
    // Traefik is configured to use the sandbox ID as the subdomain prefix.
    return `http://${port}-${this.id}.${domain}`;
  }

  // -----------------------------------------------------------------------
  // Command execution
  // -----------------------------------------------------------------------

  /**
   * Execute a command inside the sandbox and wait for the result.
   *
   * @param command - The command to execute. By default this is interpreted by
   *   the sandbox's shell; pass `mode: "argv"` to run it directly with args.
   * @param opts - Optional execution parameters.
   * @returns The execution result including exit code, stdout, stderr, and
   *   duration.
   *
   * @throws {@link SandboxNotFoundError} if the sandbox no longer exists.
   * @throws {@link ProviderError} on server-side failures.
   *
   * @example
   * ```ts
   * const result = await sandbox.exec("ls -la /tmp");
   * if (result.exit_code !== 0) {
   *   console.error(result.stderr);
   * }
   * ```
   */
  async exec(command: string, opts?: ExecOptions): Promise<ExecResult> {
    const body: Record<string, unknown> = { command };
    if (opts?.args) body["args"] = opts.args;
    if (opts?.mode) body["mode"] = opts.mode;
    if (opts?.env) body["env"] = opts.env;
    if (opts?.workdir) body["workdir"] = opts.workdir;
    if (opts?.timeout) body["timeout"] = opts.timeout;

    const response = await this._fetch(
      `/api/v1/sandboxes/${this.id}/exec`,
      {
        method: "POST",
        body: JSON.stringify(body),
      },
    );

    await handleResponse(response, this.id);
    return (await response.json()) as ExecResult;
  }

  /**
   * Execute a command and stream output chunks as they arrive.
   *
   * Returns an {@link AsyncGenerator} that yields {@link StreamChunk}
   * objects.  Each chunk contains a `stream` field (`"stdout"` or
   * `"stderr"`) and a `data` field with the text payload.
   *
   * @param command - The command string to execute.
   * @param opts - Optional execution parameters (same as {@link exec}
   *   except `timeout` is not supported for streaming).
   *
   * @yields {StreamChunk} Parsed output chunks in real time.
   *
   * @example
   * ```ts
   * for await (const chunk of sandbox.execStream("ping -c 5 localhost")) {
   *   process.stdout.write(chunk.data);
   * }
   * ```
   */
  async *execStream(
    command: string,
    opts?: Omit<ExecOptions, "timeout">,
  ): AsyncGenerator<StreamChunk, void, undefined> {
    const body: Record<string, unknown> = { command, stream: true };
    if (opts?.args) body["args"] = opts.args;
    if (opts?.mode) body["mode"] = opts.mode;
    if (opts?.env) body["env"] = opts.env;
    if (opts?.workdir) body["workdir"] = opts.workdir;

    const response = await this._fetch(
      `/api/v1/sandboxes/${this.id}/exec`,
      {
        method: "POST",
        body: JSON.stringify(body),
      },
    );

    await handleResponse(response, this.id);

    if (!response.body) {
      return;
    }

    yield* parseNDJSON(response.body);
  }

  // -----------------------------------------------------------------------
  // File operations
  // -----------------------------------------------------------------------

  /**
   * Write a file inside the sandbox.
   *
   * @param path - Absolute path inside the sandbox (e.g. `"/tmp/hello.txt"`).
   * @param content - UTF-8 string content to write.
   * @param mode - Optional Unix file mode string (e.g. `"0644"`).
   *
   * @example
   * ```ts
   * await sandbox.writeFile("/app/config.json", JSON.stringify(config));
   * ```
   */
  async writeFile(
    path: string,
    content: string,
    mode?: string,
  ): Promise<void> {
    const body: Record<string, unknown> = { path, content };
    if (mode) body["mode"] = mode;

    const response = await this._fetch(
      `/api/v1/sandboxes/${this.id}/files`,
      {
        method: "POST",
        body: JSON.stringify(body),
      },
    );

    await handleResponse(response, this.id);
  }

  /**
   * Read a file from the sandbox.
   *
   * @param path - Absolute path inside the sandbox.
   * @returns The file content as a UTF-8 string.
   *
   * @example
   * ```ts
   * const content = await sandbox.readFile("/etc/hostname");
   * ```
   */
  async readFile(path: string): Promise<string> {
    const params = new URLSearchParams({ path });
    const response = await this._fetch(
      `/api/v1/sandboxes/${this.id}/files?${params.toString()}`,
      { method: "GET" },
    );

    await handleResponse(response, this.id);
    return response.text();
  }

  /**
   * List files in a directory inside the sandbox.
   *
   * @param path - Directory path to list (default: `"/"`).
   * @returns An array of {@link FileInfo} objects.
   *
   * @example
   * ```ts
   * const files = await sandbox.listFiles("/tmp");
   * for (const f of files) {
   *   console.log(`${f.path}  ${f.size} bytes`);
   * }
   * ```
   */
  async listFiles(path: string = "/"): Promise<FileInfo[]> {
    const params = new URLSearchParams({ path });
    const response = await this._fetch(
      `/api/v1/sandboxes/${this.id}/files/list?${params.toString()}`,
      { method: "GET" },
    );

    await handleResponse(response, this.id);
    return (await response.json()) as FileInfo[];
  }

  /**
   * Delete a file or directory inside the sandbox.
   *
   * @param path - Absolute path inside the sandbox.
   * @param recursive - If true, recursively delete directories.
   */
  async deleteFile(path: string, recursive?: boolean): Promise<void> {
    const params = new URLSearchParams({ path });
    if (recursive) params.set("recursive", "true");
    const response = await this._fetch(
      `/api/v1/sandboxes/${this.id}/files?${params.toString()}`,
      { method: "DELETE" },
    );
    await handleResponse(response, this.id);
  }

  /**
   * Move/rename a file inside the sandbox.
   *
   * @param oldPath - Source path.
   * @param newPath - Destination path.
   */
  async moveFile(oldPath: string, newPath: string): Promise<void> {
    const response = await this._fetch(
      `/api/v1/sandboxes/${this.id}/files/move`,
      {
        method: "POST",
        body: JSON.stringify({ old_path: oldPath, new_path: newPath }),
      },
    );
    await handleResponse(response, this.id);
  }

  /**
   * Change file permissions inside the sandbox.
   *
   * @param path - Absolute path inside the sandbox.
   * @param mode - Octal permission string (e.g. `"0755"`).
   */
  async chmodFile(path: string, mode: string): Promise<void> {
    const response = await this._fetch(
      `/api/v1/sandboxes/${this.id}/files/chmod`,
      {
        method: "POST",
        body: JSON.stringify({ path, mode }),
      },
    );
    await handleResponse(response, this.id);
  }

  /**
   * Get file info for a single file inside the sandbox.
   *
   * @param path - Absolute path inside the sandbox.
   * @returns File info object.
   */
  async statFile(path: string): Promise<FileInfo> {
    const params = new URLSearchParams({ path });
    const response = await this._fetch(
      `/api/v1/sandboxes/${this.id}/files/stat?${params.toString()}`,
      { method: "GET" },
    );
    await handleResponse(response, this.id);
    return (await response.json()) as FileInfo;
  }

  /**
   * Return paths matching a glob pattern inside the sandbox.
   *
   * @param pattern - Glob pattern (e.g. `"/tmp/*.txt"`).
   * @returns Array of matching paths.
   */
  async globFiles(pattern: string): Promise<string[]> {
    const params = new URLSearchParams({ pattern });
    const response = await this._fetch(
      `/api/v1/sandboxes/${this.id}/files/glob?${params.toString()}`,
      { method: "GET" },
    );
    await handleResponse(response, this.id);
    return (await response.json()) as string[];
  }

  // -----------------------------------------------------------------------
  // TTL management
  // -----------------------------------------------------------------------

  /**
   * Extend this sandbox's TTL by the given duration.
   *
   * @param ttl - A Go-style duration string (e.g. `"30m"`, `"1h"`).
   *   The new expiry is calculated from now, not from the current expiry.
   *
   * @example
   * ```ts
   * await sandbox.extendTtl("30m"); // extends by 30 minutes
   * ```
   */
  async extendTtl(ttl: string = "30m"): Promise<void> {
    const response = await this._fetch(
      `/api/v1/sandboxes/${this.id}/extend`,
      {
        method: "POST",
        body: JSON.stringify({ ttl }),
      },
    );

    await handleResponse(response, this.id);
    this._info = (await response.json()) as SandboxInfo;
  }

  // -----------------------------------------------------------------------
  // Lifecycle
  // -----------------------------------------------------------------------

  /**
   * Destroy this sandbox, releasing all resources.
   *
   * After calling `destroy()` no further operations should be performed on
   * this {@link Sandbox} instance.
   *
   * @throws {@link SandboxNotFoundError} if the sandbox was already
   *   destroyed.
   */
  async destroy(): Promise<void> {
    const response = await this._fetch(
      `/api/v1/sandboxes/${this.id}`,
      { method: "DELETE" },
    );

    await handleResponse(response, this.id);
    this._info = { ...this._info, state: "destroyed" };
  }

  /**
   * Refresh the cached sandbox info from the server.
   *
   * Call this to get the latest state if you suspect the sandbox may have
   * changed (e.g. expired via TTL).
   *
   * @returns This {@link Sandbox} instance (for chaining).
   */
  async refresh(): Promise<Sandbox> {
    const response = await this._fetch(
      `/api/v1/sandboxes/${this.id}`,
      { method: "GET" },
    );

    await handleResponse(response, this.id);
    this._info = (await response.json()) as SandboxInfo;
    return this;
  }

  // -----------------------------------------------------------------------
  // Internal helpers
  // -----------------------------------------------------------------------

  /**
   * Issue an HTTP request against the StacyVM API.
   *
   * Handles base URL construction, default headers, and timeout via
   * `AbortSignal.timeout`.
   */
  private async _fetch(
    path: string,
    init: RequestInit,
  ): Promise<Response> {
    const url = `${this._baseUrl}${path}`;
    const headers: Record<string, string> = {
      ...this._headers,
    };

    // Only set Content-Type for requests that have a body.
    if (init.body) {
      headers["Content-Type"] = "application/json";
    }

    return fetch(url, {
      ...init,
      headers: { ...headers, ...(init.headers as Record<string, string>) },
      signal: AbortSignal.timeout(this._timeout),
    });
  }

  // -----------------------------------------------------------------------
  // Debug helpers
  // -----------------------------------------------------------------------

  /** Returns a human-readable string representation. */
  toString(): string {
    return `Sandbox(id="${this.id}", state="${this.state}")`;
  }

  /** Customise `JSON.stringify` output. */
  toJSON(): SandboxInfo {
    return this.info;
  }
}
